### PR TITLE
V1.2

### PR DESCRIPTION
**Pull request by olivercheok20**

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddPolicyTagCommand.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                policyToEdit.getName(),
+                policyToEdit.getDescription(),
+                policyToEdit.getCoverage(),
+                policyToEdit.getPrice(),
+                policyToEdit.getStartAge(),
+                policyToEdit.getEndAge(),
+                policyToEdit.getCriteria(),
+                newTags
+        );
+
```
Comment (chaitanyabaranwal) [inline src/main/java/seedu/address/logic/commands/AddPolicyTagCommand.java]: You can use the `PersonBuilder` class to build this.

File: src/main/java/seedu/address/logic/commands/AddTagCommand.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                personToEdit.getName(),
+                personToEdit.getNric(),
+                personToEdit.getPhone(),
+                personToEdit.getEmail(),
+                personToEdit.getAddress(),
+                personToEdit.getDateOfBirth(),
+                personToEdit.getPolicies(),
+                newTags
+        );
+
```
Comment (chaitanyabaranwal) [inline src/main/java/seedu/address/logic/commands/AddTagCommand.java]: Can use `PersonBuilder` here also

File: src/main/java/seedu/address/logic/commands/DeletePolicyTagCommand.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                policyToEdit.getName(),
+                policyToEdit.getDescription(),
+                policyToEdit.getCoverage(),
+                policyToEdit.getPrice(),
+                policyToEdit.getStartAge(),
+                policyToEdit.getEndAge(),
+                policyToEdit.getCriteria(),
+                newTags
+        );
+
```
Comment (chaitanyabaranwal) [inline src/main/java/seedu/address/logic/commands/DeletePolicyTagCommand.java]: Can use `PolicyBuilder` here

File: src/main/java/seedu/address/logic/commands/DeleteTagCommand.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        Person editedPerson = new Person(
+                personToEdit.getName(),
+                personToEdit.getNric(),
+                personToEdit.getPhone(),
+                personToEdit.getEmail(),
+                personToEdit.getAddress(),
+                personToEdit.getDateOfBirth(),
+                personToEdit.getPolicies(),
+                newTags
+        );
```
Comment (chaitanyabaranwal) [inline src/main/java/seedu/address/logic/commands/DeleteTagCommand.java]: Use `PersonBuilder` here

**Comment by chaitanyabaranwal** [comment]
Can use `PersonBuilder` and `PolicyBuilder` to copy people/policies!

File: src/main/java/seedu/address/logic/commands/DeletePolicyTagCommand.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        model.setPolicy(policyToEdit, editedPolicy);
+        model.updateFilteredPolicyList(PREDICATE_SHOW_ALL_POLICIES);
```
Comment (chowyiyin) [inline src/main/java/seedu/address/logic/commands/DeletePolicyTagCommand.java]: Maybe you can add in the part where you update the policies that are already assigned to a person?

Comment (chaitanyabaranwal) [reply src/main/java/seedu/address/logic/commands/DeletePolicyTagCommand.java]: Yep! You can check `editPolicy` command to see how it's being done right now.

File: src/main/java/seedu/address/logic/commands/AddPolicyTagCommand.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        model.setPolicy(policyToEdit, editedPolicy);
+        model.updateFilteredPolicyList(PREDICATE_SHOW_ALL_POLICIES);
```
Comment (chaitanyabaranwal) [inline src/main/java/seedu/address/logic/commands/AddPolicyTagCommand.java]: Same thing goes here, look at `editPolicy` to see how we are updating people who are assigned those policies.

File: src/main/java/seedu/address/logic/commands/DeletePolicyTagCommand.java (hunk 5 above)
Comment (olivercheok20) [reply src/main/java/seedu/address/logic/commands/DeletePolicyTagCommand.java]: Oh okay sure thing! My bad I didn't realise about that haha

File: src/main/java/seedu/address/logic/commands/AddPolicyTagCommand.java (hunk 6 above)
Comment (olivercheok20) [reply src/main/java/seedu/address/logic/commands/AddPolicyTagCommand.java]: Fixed in latest commit!

**Comment by chowyiyin** [comment]
> Hi Oliver, I tested to check if everything works fine. Everything is okay except:
> When you type commands like addtag 1 or addpolicytag 1, the command executes without giving an error, even when it should give an exception that no tags are provided (along with a correct format for the command).

**Comment by chowyiyin** [comment]
You can just add this to your addpolicytagcommandparser:
if (!arePrefixesPresent(argMultimap, PREFIX_TAG)
                || !argMultimap.getPreamble().isEmpty()) {
            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddPolicyTagCommand.MESSAGE_USAGE));
        }

File: src/main/java/seedu/address/logic/commands/DeletePolicyTagCommand.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.Model;
+import seedu.address.model.person.Person;
+import seedu.address.model.policy.Policy;
+import seedu.address.model.tag.Tag;
+
+/**
+ * Deletes a tag of an existing policy in the address book.
+ */
+public class DeletePolicyTagCommand extends Command {
```
Comment (larrylawl) [inline src/main/java/seedu/address/logic/commands/DeletePolicyTagCommand.java]: Bug: success message shown when I delete a non-existent tag

File: src/main/java/seedu/address/logic/commands/DeleteTagCommand.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+import seedu.address.commons.util.PersonBuilder;
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.Model;
+import seedu.address.model.person.Person;
+import seedu.address.model.tag.Tag;
+
+/**
+ * Deletes a tag of an existing person in the address book.
+ */
+public class DeleteTagCommand extends Command {
```
Comment (larrylawl) [inline src/main/java/seedu/address/logic/commands/DeleteTagCommand.java]: Same bug mentioned above for delete tag

File: src/main/java/seedu/address/logic/commands/FindPolicyCommand.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+import seedu.address.commons.core.Messages;
+import seedu.address.model.Model;
+import seedu.address.model.policy.PolicyNameContainsKeywordsPredicate;
+
+/**
+ * Finds and lists all policies in address book whose name contains any of the argument keywords.
+ * Keyword matching is case insensitive.
+ */
+public class FindPolicyCommand extends Command {
```
Comment (larrylawl) [inline src/main/java/seedu/address/logic/commands/FindPolicyCommand.java]: Bug: Using `findpolicy` while the list is showing people still shows list of people.

Comment (olivercheok20) [reply src/main/java/seedu/address/logic/commands/FindPolicyCommand.java]: Will work on this, thank you! Seems to be a bug with `find` as well: using `find` while list is showing policies still shows list of policies.

**Pull request merged**